### ISSUE TITLE
Improve tools for making an auto-generated post

### ISSUE THREAD

**Issue by akhmerov** [getnikola/nikola#1553]
Right now auto-generated posts or pages require two executions of `nikola build`. The reason for this is that scanning for posts is out of user control. So since the scan is done before the tasks are generated, the newly added files are missed. While the inconvenience of running build twice is small, a more robust machinery would still be good to have. Auto-generated posts are created by planetoid, and I needed an auto-generated page to create a publication list. For attempts to make a solution and more details see https://groups.google.com/forum/#!topic/nikola-discuss/J9k1sCM8VO0

**Comment by felixfontein** [comment]
I think there's a problem with the current implementation of ```EarlyTask```s with this use case. Namely, while the ```EarlyTask```s ```gen_task()``` function will be called before regular ```Task```s (and thus before ```Nikola.scan_posts()```, which is called in ```RenderPosts.gen_task()```), but the tasks created by ```EarlyTask``` plugins are *not* built before ```Nikola.scan_posts()``` is run. Therefore, if an ```EarlyTask``` plugin generates posts, they are not found during that run of Nikola.

**Comment by Kwpolska** [comment]
How else do you recommend to implement it?

Worst case scenario, earlytasks can rescan as part of the procedure.

**Comment by felixfontein** [comment]
Well, I have no idea. I think it's impossible without modifying/extending doit. If doit would run a function ```get_more_tasks``` every time it runs out of tasks and continues until that function returns an empty list (or something like that), it would be possible to properly implement this.
Rescanning doesn't really help since the scan is needed to create the tasks for processing the posts, and that must be done (with current doit) before any task is executed. (At least it won't work without some hacks, and probably only ugly ones.)

**Comment by Kwpolska** [comment]
@schettino72, do you have any ideas?

**Comment by ralsina** [comment]
A very different solution is to use a Command plugin to generate the files.
So, suppose the files are reports:

````
nikola gen_reports -foo -bar
nikola build
````

Creates a clear separation between "this action generates files" and "this action builds the site".
If you actually always need to run both (which is strange), do a script.

**Comment by akhmerov** [comment]
@ralsina a `Command` doesn't resolve the issue. After all just running `nikola build` twice will have the same effect.

**Comment by ralsina** [comment]
@akhmerov it is different semantically. Usually we try to make builds something that can succeed in one step (although I have some ideas about multi-pass builds, which are needed for some things)

**Comment by akhmerov** [comment]
Well, right, and in that sense introducing a `Command` doesn't fix the issue of a build not succeeding in one step if the posts/pages are not known in advance.

**Comment by felixfontein** [comment]
(One thing which works is not using doit to determine whether the autogenerated post needs to be created or not, but to simply do that yourself (or generate it everytime) in the plugin's ```set_site``` method. Then it will be there when the posts are scanned. It's kind of a hack, but works without any change to Nikola.)

**Comment by ralsina** [comment]
@akhmerov yes, but separating it into a step that is clearly separated (pardon the redundancy) actually moves the problem around. *You* need to decide whether the posts need creating or not, and then for Nikola it's business as usual.

I'll spend some time later today thinking of a more general solution tho :-)

**Comment by schettino72** [comment]
> Auto-generated posts are created by planetoid,

planetoid? thats your plugin? where is the code for this?
Can you explain how it creates new posts? how many? where the name of the posts come from, etc... The best/easiest solution depends on that.

From the original email I understood that the auto-generated page is specified in a config.
Is that correct?

### If you know which posts will be generated at task-creation time

Currently the only way nikola finds posts is by scanning the file system to find source of posts.
Nikola will also need to be configured to create posts tasks from other sources (like a list taken from your config). Then as long as your plugin includes the file name of the generated post in "targets" doit will make sure it is executed in the right order. 

### if you really dont know which (how many) posts will be auto-generated

If this is the case doit will need to be modified to handle it. 2 options:

1) allow creation of new tasks at run-time by other tasks. IMO creating more tasks at task-execution time should be avoided if possible but in some cases there is no other way.

2) delay the creation of certain tasks. i.e. create posts tasks only after the auto-generate posts tasks is executed. Actually the reason I designed doit to create task metadata from a function (and not directly in the module scope) was to make this possible, but i never needed/implemented that.

**Comment by ralsina** [comment]
I wrote planetoid, and it works, mostly, for what I wanted it to do, I expect his plugin is something else :)

In fact, different pieces of Nikola do create Post objects, for example, the gallery plugin does it to create the index pages. They are more like stories in that they are not in the timeline, and it doesn't seem to be a problem.

**Comment by felixfontein** [comment]
@schettino72: the very last thing you wrote ("delay the creation of certain tasks") is essentially what I wanted to describe in #1562. I think something like that would be the cleanest solution to the problem (or would allow to implement it).

**Comment by felixfontein** [comment]
@akhmerov: if you want you can try the ```earlytask_impl``` branch of Nikola. It provides a plugin type ```EarlyTask``` which is always executed _before_ posts are scanned.

It is still experimental, but seems to work fine at the moment (as long as you don't do something not supported, which is mainly specifying task/target names after ```nikola build```). Some more details are in #1562.

**Comment by felixfontein** [comment]
(Forgot to mention: you need doit master from github to run this. The current doit release won't do it.)

**Comment by akhmerov** [comment]
@felixfontein Thanks, I verified that this branch works as intended with my plugin.

**Comment by felixfontein** [comment]
Cool! That's good news :)

**Comment by felixfontein** [comment]
This can be solved by creating a post scanning plugin which first writes the generated posts into some subdirectory of `cache/` and then provides the corresponding `Post` objects as a result of its scanning.

(With #2659 the post files don't even have to be written somewhere on disk, but it is not clear whether this will ever come.)

**Comment by Kwpolska** [comment]
EarlyTasks probably won’t happen, and the solution with the custom post scanner seems reasonable, so I’m closing this.
